### DISCUSSION TITLE
Feature request: Support for postgresql's (NOTIFY & )LISTEN

### DISCUSSION THREAD

**Discussion by timwoocker** [sqlalchemy/sqlalchemy#11957] [Ideas]
I'm currently using [`NOTIFY`](https://www.postgresql.org/docs/current/sql-notify.html) in a db trigger.
Handling events from that notify in python currently requires me to use psycopg (`notifies` generator) or asyncpg (`add_listener` method) directly because SQLAlchemy doesn't provide any methods to use [`LISTEN`](https://www.postgresql.org/docs/current/sql-listen.html) (or more precisely iterate over the notifies).

That means when using SQLAlchemy, I need to maintain separate connections from psycopg or asyncpg or access the raw connection using `(await conn.get_raw_connection()).driver_connection`.
Now I rely on specifically using asyncpg (in my case) - switching to psycopg as the driver would require me to rewrite my code.
Also I need to catch specific asyncpg exceptions and the handler always gets passed the asyncpg connection instead of the SQLAlchemy connection.

I haven't found any other discussions about this feature here yet so I figured I'd open it myself.
This feature only applies to postgres drivers (idk if other dbms have similar features) and it is possible to do with a few detours, so I would understand if this is not something SQLAlchemy should focus on.
For me personally however, it would make things much easier.

Here is a simplified version of how I'm currently handling it:

```python
async def handle_notify(_, __, channel, payload):
    print(f"Got NOTIFY in {channel} with payload {payload}")
    ...

while True:
    close_signal = asyncio.Event()

    try:
        async with engine.connect() as conn:
            raw_conn: asyncpg.Connection = (await conn.get_raw_connection()).driver_connection
            raw_conn.add_termination_listener(lambda *_, *__: close_signal.set())
            await raw_conn.add_listener("channelnameyeah", handle_notify)
            print("Listening.")
            await close_signal.wait()
            await conn.invalidate()
            print("DB connection closed. Reconnecting...")
    ...
```

**Comment by zzzeek** [comment]
hi -

if you have asyncpg code that does this, I woudn't consider adding support for psycopg a "rewrite", you'd be just adding the implementation for more drivers.    this is work that would have to be done if SQLAlchemy were to have such a "feature" in any case.

a SQLAlchemy side feature would not normally be using async primitives since we have an [event system](https://docs.sqlalchemy.org/en/20/core/event.html#events) which is what we use for all callbacks, since if we included "notify" support it would need to work for most or all PostgreSQL dialects and certainly not just the async ones in any case.    you can use events with asyncio code as described [here](https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#using-events-with-the-asyncio-extension).

to generalize this would mean we'd need to identify the "notify" feature in each PG DBAPI and build an event handler that encompasses the common featureset of each DBAPI in a consistent way.

there's nothing too controversial here, just need to design an API appropriately and it would also need tests and documentation.
